Postinstall script installation of tabtab fails

<!--
1. If you have a question and not a bug/feature request please ask it at http://forum.serverless.com
2. Please check if an issue already exists so there are no duplicates
3. Check out and follow our Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
4. Fill out the whole template so we have a good overview on the issue
5. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the Issue
6. Please follow the template, otherwise we'll have to ask you to update it
-->

# This is a Bug Report

## Description

When installing `serverless` the `postinstall` script throws when attempting to run `tabtab install` because `tabtab` does not exist. If I run the `tamtam install` command separately after installation, it succeeds. There seems to be a type of race condition with the creation of that binary and the execution of the postinstall script.

Repro steps:
1. Run `npm install serverless`
2. Observer error message on completion

Node.js version: 6.10.3 (also tried with 8.0.0)
npm version: 5.0.3

## Additional Data

* ***Serverless Framework Version you're using***: 1.15.1
* ***Operating System***: MacOS 10.12.4
* ***Stack Trace***: 
```sh
> serverless@1.15.1 postinstall /Users/rylewis/Documents/Dev/Repo/deploy/node_modules/serverless
> node ./scripts/postinstall.js

/bin/sh: ./node_modules/tabtab/bin/tabtab: No such file or directory
/bin/sh: ./node_modules/tabtab/bin/tabtab: No such file or directory
Unhandled rejection Error: Command failed: ./node_modules/tabtab/bin/tabtab install --name serverless --stdout
/bin/sh: ./node_modules/tabtab/bin/tabtab: No such file or directory

    at checkExecSyncError (child_process.js:481:13)
    at execSync (child_process.js:521:13)
    at serverless.init.then.e (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/serverless/scripts/postinstall.js:15:7)
    at tryCatcher (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:693:18)
    at Promise._fulfill (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:638:18)
    at Promise._settlePromise (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:582:21)
    at Promise._settlePromise0 (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:693:18)
    at Promise._fulfill (/Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/promise.js:638:18)
    at /Users/rylewis/Documents/Dev/Repo/deploy/node_modules/bluebird/js/release/nodeback.js:42:21
    at /Users/rylewis/Documents/Dev/Repo/deploy/node_modules/graceful-fs/graceful-fs.js:99:16
    at /Users/rylewis/Documents/Dev/Repo/deploy/node_modules/graceful-fs/graceful-fs.js:43:10
    at FSReqWrap.oncomplete (fs.js:123:15)
```

We are also getting an error when installing 1.15.1, but the error looks different.

```
/usr/lib/node_modules/serverless/node_modules/write-file-atomic/index.js:132
    throw err
    ^

Error: EACCES: permission denied, open '/root/.serverlessrc.1718049435'
    at Error (native)
    at Object.fs.openSync (fs.js:641:18)
    at Function.writeFileSync [as sync] (/usr/lib/node_modules/serverless/node_modules/write-file-atomic/index.js:117:17)
    at createConfig (/usr/lib/node_modules/serverless/lib/utils/config/index.js:30:19)
    at getConfig (/usr/lib/node_modules/serverless/lib/utils/config/index.js:44:5)
    at Object.get (/usr/lib/node_modules/serverless/lib/utils/config/index.js:90:18)
    at isTrackingDisabled (/usr/lib/node_modules/serverless/lib/utils/isTrackingDisabled.js:6:22)
    at Object.<anonymous> (/usr/lib/node_modules/serverless/lib/utils/segment.js:9:30)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/usr/lib/node_modules/serverless/lib/classes/Utils.js:14:17)
npm ERR! Linux 4.9.27-moby
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "install" "-g" "serverless"
npm ERR! node v6.10.3
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE

npm ERR! serverless@1.15.1 postinstall: `node ./scripts/postinstall.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the serverless@1.15.1 postinstall script 'node ./scripts/postinstall.js'.
```

Environment:
Centos 7
node 6.10.2
npm 3.10.10

Is this the same problem, or something different?

--- a/lib/Serverless.js
+++ b/lib/Serverless.js
@@ -16,11 +16,11 @@ const ServerlessError = require('./classes/Error').ServerlessError;
 const Version = require('./../package.json').version;
 const initializeErrorReporter = require('./utils/sentry').initializeErrorReporter;
 
-// boot up error reporting via sentry
-initializeErrorReporter();
-
 class Serverless {
   constructor(config) {
+    // boot up error reporting via sentry
+    initializeErrorReporter();
+
     let configObject = config;
     configObject = configObject || {};
 

--- a/lib/utils/segment.js
+++ b/lib/utils/segment.js
@@ -3,10 +3,10 @@
 const BbPromise = require('bluebird');
 const fetch = require('node-fetch');
 const isTrackingDisabled = require('./isTrackingDisabled');
+
 // TODO: Replace me before release
 const writeKey = 'XXXX';
 const auth = `${writeKey}:`;
-const TRACKING_IS_DISABLED = isTrackingDisabled();
 
 /* note segment call swallows errors */
 function request(url, payload) {
@@ -25,6 +25,7 @@ function request(url, payload) {
 }
 
 function track(payload) {
+  const TRACKING_IS_DISABLED = isTrackingDisabled();
   // exit early is tracking disabled
   if (TRACKING_IS_DISABLED) {
     return BbPromise.resolve();

--- a/lib/utils/sentry.js
+++ b/lib/utils/sentry.js
@@ -7,10 +7,10 @@ const DSN = 'https://cbb3655b343a49ee9a18494d0dd171a7:4b9ef9a2c5eb40379f30b5e680
 const SLS_DISABLE_ERROR_TRACKING = true;
 const IS_CI = ci.isCI;
 
-
 function initializeErrorReporter() {
   const config = configUtils.getConfig();
   const trackingDisabled = config.trackingDisabled;
+
   // exit if tracking disabled or inside CI system
   if (SLS_DISABLE_ERROR_TRACKING || trackingDisabled || IS_CI) {
     // console.log('tracking off')

--- a/lib/utils/userStats.js
+++ b/lib/utils/userStats.js
@@ -9,7 +9,6 @@ const configUtils = require('./config');
 const isTrackingDisabled = require('./isTrackingDisabled');
 const isValidEventName = require('./userStatsValidation');
 
-const TRACKING_IS_DISABLED = isTrackingDisabled();
 const TRACK_URL = 'https://serverless.com/api/framework/track';
 const IDENTIFY_URL = 'https://serverless.com/api/framework/identify';
 const DEBUG = false;
@@ -40,6 +39,8 @@ function track(eventName, payload) {
   const data = payload || {};
   let userId = data.id;
   let userEmail = data.email;
+
+  const TRACKING_IS_DISABLED = isTrackingDisabled();
 
   // exit early if tracking disabled
   if (TRACKING_IS_DISABLED && !data.force) {
@@ -89,7 +90,9 @@ function track(eventName, payload) {
 }
 
 function identify(payload) {
+  const TRACKING_IS_DISABLED = isTrackingDisabled();
   const data = payload || {};
+
   if (TRACKING_IS_DISABLED && !data.force) {
     if (DEBUG) {
       console.log('abort .identify call');

--- a/scripts/postinstall.js
+++ b/scripts/postinstall.js
@@ -1,22 +1,35 @@
 'use strict';
 
-const execSync = require('child_process').execSync;
+/* eslint-disable no-console */
+/* eslint-disable no-use-before-define */
 
 const Serverless = require('../lib/Serverless');
+const execSync = require('child_process').execSync;
 
-const serverless = new Serverless();
+try {
+  const serverless = new Serverless();
 
-(() => {
-  serverless.init().then(() => {
+  (() => serverless.init()
+    .then(() => serverless.utils.logStat(serverless, 'install'))
+    .then(() => setupAutocomplete())
+    .catch(() => Promise.resolve())
+  )();
+} catch (error) {
+  // fail silently
+}
+
+function setupAutocomplete() {
+  return new Promise((resolve, reject) => {
     try {
-      execSync('./node_modules/tabtab/bin/tabtab install --name serverless --auto');
-      execSync('./node_modules/tabtab/bin/tabtab install --name sls --auto');
-    } catch (e) {
-      execSync('./node_modules/tabtab/bin/tabtab install --name serverless --stdout');
-      execSync('./node_modules/tabtab/bin/tabtab install --name sls --stdout');
-      serverless.cli.consoleLog('Could not auto-install serverless autocomplete script.');
-      serverless.cli.consoleLog('Please copy/paste the script above into your shell.');
+      execSync('node ./node_modules/tabtab/src/cli.js install --name serverless --auto');
+      execSync('node ./node_modules/tabtab/src/cli.js install --name sls --auto');
+      return resolve();
+    } catch (error) {
+      execSync('node ./node_modules/tabtab/src/cli.js install --name serverless --stdout');
+      execSync('node ./node_modules/tabtab/src/cli.js install --name sls --stdout');
+      console.log('Could not auto-install serverless autocomplete script.');
+      console.log('Please copy / paste the script above into your shell.');
+      return reject(error);
     }
-    return serverless.utils.logStat(serverless, 'install').catch(() => Promise.resolve());
   });
-})();
+}

--- a/scripts/preuninstall.js
+++ b/scripts/preuninstall.js
@@ -2,10 +2,13 @@
 
 const Serverless = require('../lib/Serverless');
 
-const serverless = new Serverless();
+try {
+  const serverless = new Serverless();
 
-(() => {
-  serverless.init().then(() => {
-    serverless.utils.logStat(serverless, 'uninstall').catch(() => Promise.resolve());
-  });
-})();
+  (() => serverless.init()
+    .then(() => serverless.utils.logStat(serverless, 'uninstall'))
+    .catch(() => Promise.resolve())
+  )();
+} catch (error) {
+  // fail silently
+}

